lib/ecto_shorts/actions.ex

```diff
@@ -0,0 +1,857 @@
+    query
+    |> CommonSchemas.get_schema_query()
+    |> CommonFilters.convert_params_to_filter(params)
+    |> replica!(opts).aggregate(aggregate, field, opts)
+  end
+
+  @doc """
+  Accepts a list of schemas and attempts to find them in the DB. Any missing Schemas will be created.
+  Can also accept a keyword options list.
+
+  ***Note: Relational filtering doesn't work on this function***
+
+  ## Options
+    * `:repo` - A module that uses the Ecto.Repo Module.
+    * `:replica` - If you don't want to perform any reads against your Primary, you can specify a replica to read from.
+
+  ## Examples
+    iex> {:ok, records} = EctoSchemas.Actions.find_or_create_many(EctoSchemas.Accounts.User, [%{name: "foo"}, %{name: "bar}])
+    iex> length(records) === 2
+  """
+  @spec find_or_create_many(
+    query :: queryable() | source_queryable(),
+    params_list :: params_list(),
+    opts :: opts()
+  ) :: {:ok, schemas()} | {:error, changesets()}
+  @spec find_or_create_many(
+    query :: queryable() | source_queryable(),
+    params_list :: params_list()
+  ) :: {:ok, schemas()} | {:error, changesets()}
+  def find_or_create_many(query, param_list, opts) do
+    queryable = CommonSchemas.get_schema_queryable(query)
+
+    find_param_list =
+      Enum.map(param_list, fn params ->
+        drop_associations(params, queryable)
+      end)
+
+    {create_params, found_results} = find_many(query, find_param_list, opts)
+
+    query
+    |> multi_insert(param_list, create_params, opts)
+    |> repo!(opts).transaction()
+    |> case do
+      {:ok, created_map} -> {:ok, merge_found(created_map, found_results)}
+      error -> error
+    end
+  end
+
+  def find_or_create_many(schema, param_list) do
+    find_or_create_many(schema, param_list, default_opts())
+  end
+
+  defp find_many(schema, param_list, opts) do
+    param_list
+    |> Enum.map(fn params ->
+      case find(schema, params, opts) do
+        {:ok, schema_data} -> schema_data
+        _ -> nil
+      end
+    end)
+    |> Enum.with_index()
+    |> Enum.split_with(fn {schema_data, _index} -> is_nil(schema_data) end)
+  end
+
+  defp multi_insert(queryable, param_list, create_params, opts) do
+    Enum.reduce(create_params, Ecto.Multi.new(), fn {nil, i}, multi ->
+      Ecto.Multi.insert(multi, i, fn _ ->
+        build_changeset(queryable, Enum.at(param_list, i), opts)
+      end)
+    end)
+  end
+
+  defp build_changeset({source, queryable}, schema_data, params, opts) do
+    schema_data = SchemaHelpers.build_struct(schema_data, source: source)
+
+    build_changeset(queryable, schema_data, params, opts)
+  end
+
+  defp build_changeset(queryable, schema_data, params, opts) do
+    case opts[:changeset] do
+      nil ->
+        queryable.changeset(schema_data, params)
+
+      func when is_function(func, 2) ->
+        func.(schema_data, params)
+
+      func when is_function(func, 1) ->
+        schema_data
+        |> queryable.changeset(params)
+        |> func.()
+    end
+  end
```

Comment:
When using an abstract schema the "assoc_id" is a polymorphic reference and the name of the foreign key is different per table. This function was added so that constraints can be added to those schemas with the appropriate constraint names before the changeset is handed off to the `Ecto.Repo` api.

Reply:
https://github.com/MikaAK/ecto_shorts/blob/386a293727eae3c0f4d287d44c912b0ac0f04acd/test/ecto_shorts/actions_abstract_schema_test.exs#L117